activesupport/lib/active_support/messages/rotation_coordinator.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module Messages
+    class RotationCoordinator # :nodoc:
+      def initialize(&secret_generator)
+        raise ArgumentError, "A secret generator block is required" unless secret_generator
+        @secret_generator = secret_generator
+        @rotate_options = []
+        @codecs = {}
+      end
+
+      def [](salt)
+        @codecs[salt] ||= build_with_rotations(salt)
+      end
+
+      def []=(salt, codec)
+        @codecs[salt] = codec
+      end
+
+      def rotate(**options)
+        changing_configuration!
+        @rotate_options << { secret_generator: @secret_generator, **options }
+        self
+      end
+
+      def rotate_defaults
+        rotate()
+      end
+
+      def clear_rotations
+        changing_configuration!
+        @rotate_options.clear
+        self
+      end
+
+      def on_rotation(&callback)
+        changing_configuration!
+        @on_rotation = callback
+      end
+
+      private
+        def changing_configuration!
+          if @codecs.any?
+            raise <<~MESSAGE
+              Cannot change #{self.class} configuration after it has already been applied.
+
+              The configuration has been applied with the following salts:
+              #{@codecs.keys.map { |salt| "- #{salt.inspect}" }.join("\n")}
+            MESSAGE
+          end
+        end
```

Comment:
I also wanted to point out another design decision that I made, in case there are any objections: rotations cannot be added nor cleared after a `MessageVerifier` (or `MessageEncryptor`) has been built or set via `[]=`.

The workflow I envisioned was that apps add rotations to `Rails.application.message_verifiers` in e.g. `before_initialize`, and libraries get verifiers from `Rails.application.message_verifiers` at a later stage.

The only way we could "clear rotations" after instances have been created would be to drop those instances.  But that does not help if libraries already hold references to those instances.

And thus, disallowing `rotate` is consistent with disallowing `clear_rotations`.  But also, if we allowed `rotate`, we would need to decide whether new rotations apply to instances set via `[]=`.  It feels like they should not apply, because `[]=` is an explicit override mechanism, but the inconsistency might be confusing.

The current implementation is conservative, and doesn't prevent us from expanding the behavior in the future.